cdap-common/src/main/java/io/cdap/cdap/common/metrics/MetricsReporterHook.java

```diff
@@ -69,6 +76,7 @@ public boolean preCall(HttpRequest request, HttpResponder responder, HandlerInfo
     try {
       MetricsContext collector = collectorCache.get(createContext(handlerInfo));
       collector.increment("request.received", 1);
+      request.headers().add(CDAP_REQ_TIMESTAMP_HDR, System.currentTimeMillis());
```

Comment:
I have some concern about updating the http header to carry timestamp like this, as the header can be changed by the handler method as well. It maybe better to keep a Identity map using the HttpRequest as the key.

Reply:
Also, for duration tracking purpose, the `System.nanoTime` instead of `System.currentTimeMillis()`.

Reply:
I was using a map earlier, when @tivv suggested I can use the headers.

Reply:
We already do headers enrichment in io.cdap.cdap.gateway.router.handlers.AuthenticationHandler.
But if we want to skip it, we can go Map way. But we need to ensure it's some kind of bounded map (weak key map or limited size cache) to prevent leaks if for whatever reason (e.g. client closed connection before we started to answer) postCall is not called

Reply:
AuthenticationHandler is meant to modify the request as it is a ChannelInboundHandler, and also the headers it modifies are for downstream handlers to consume for authentication purpose. The MetricsReporterHook here is more like an inspector. However, the line is blurry. We currently don't have any agreement nor central place to define these internally inserted headers. It might be good to at least put them in once place to have better understanding what are them to avoid future conflicts.

Reply:
Sure. Moved the header name to a new class called CDAPHeaderNames.